### DISCUSSION TITLE
multiple ipsec-tunnels are being negotiated between same src and dst ipaddresss

### DISCUSSION THREAD

**Discussion by trineshreddy** [strongswan/strongswan#1524] [Q&A]
hi ,
i have  2 nodes lets name them as node1 and node2
strogswan version in node1 = 5.9.8. and node 2 = 5.6.2
i am trying to establish ipsec tunnel between 2nodes.

when `swanctl --load-all` is executed  on both nodes  ipsec tunnel is established between 2 nodes successfully

but issue is after sometime nearly 20min, so many tunnels are being established between 2 nodes for same src and dst ipaddress as below ,

`# ip x  s l`

> src 11.5.213.128 dst 11.28.205.128
>         proto esp spi 0xc66d6236 reqid 1 mode tunnel
>         replay-window 0 flag af-unspec
>         aead rfc4106(gcm(aes)) 0x12e223b0e4fd86073148ddd96ccb6302b12f16d4112be9e36e617a90072ca082a8e55e12 64
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

> src 11.28.205.128 dst 11.5.213.128
>         proto esp spi 0xcb0756de reqid 1 mode tunnel
>         replay-window 32 flag af-unspec
>         aead rfc4106(gcm(aes)) 0x30ab31b5b60480e6603d22f2986ba847e4f5f784ac8fe5a121a3916002a77546898a3b77 64
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

> src 11.5.213.128 dst 11.28.205.128
>         proto esp spi 0xc923a403 reqid 1 mode tunnel
>         replay-window 0 flag af-unspec
>         aead rfc4106(gcm(aes)) 0x1e9fc9a3e6d855d69168619d9285cc191d8ef8d888488f10bc6a0853010d3ef98885e6ad 64
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

> src 11.28.205.128 dst 11.5.213.128
>         proto esp spi 0xcd8e2849 reqid 1 mode tunnel
>         replay-window 32 flag af-unspec
>         aead rfc4106(gcm(aes)) 0xcd6c5f49984ab0f2fc2878bc877650f75438d049cab83e247506cf3ecc777dbfa3bf649f 64
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

> src 11.5.213.128 dst 11.28.205.128
>         proto esp spi 0xcc4f97e4 reqid 1 mode tunnel
>         replay-window 0 flag af-unspec
>         aead rfc4106(gcm(aes)) 0x4819a9a279d50419876ed72db85ca89cd262cd6c244066bb365ecb095b1d0583e3c951a5 64
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000

> src 11.28.205.128 dst 11.5.213.128
>         proto esp spi 0xcd984616 reqid 1 mode tunnel
>         replay-window 32 flag af-unspec
>         aead rfc4106(gcm(aes)) 0xb275d845c8fca8d3c01cc36440d53609ce3bbb6599e12344b4971231992362bcc231d0ea 64
>         anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000 both the nodes are in start mode

above shown tunnels are not all tunnels established between 2 nodes.
here you can see the output of command` ip xfrm state list` in node1 which will list all ipsec tunnels
https://pastebin.pl/view/77b86280

please help in understanding why this type of behaviour may be happening?

the interfaces between which ipsec-tunnels are established are as below
1) In node1 the interface to which ipsec tunnel is created is called ipsec0 with ipaddress = 11.5.213.128/32

>     ipsec0@ctrl0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
>     link/ether fe:ff:0b:05:d5:80 brd ff:ff:ff:ff:ff:ff
>     inet 11.5.213.128/32 scope global ipsec0
>        valid_lft forever preferred_lft forever
>     inet6 fe80::fcff:bff:fe05:d580/64 scope link
>        valid_lft forever preferred_lft forever

here ipsec0 interface is created as MACVLAN interface from interface ctrl0 using below cmd
`ip link add ipsec0 link ctrl0 type macvlan mode bridge`

2)In node2 the interface to which ipsec tunnel is created is also called ipsec0 with ipaddress = 11.28.205.128/24

>    ipsec0@ipsec0-ph: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default qlen 1000
>     link/ether fe:ff:0b:1c:cd:80 brd ff:ff:ff:ff:ff:ff
>     inet 11.28.205.128/24 brd 11.28.205.255 scope global ipsec0
>        valid_lft forever preferred_lft forever
>     inet6 fe80::fcff:bff:fe1c:cd80/64 scope link 
>        valid_lft forever preferred_lft forever

        
### **useful information in node1:**

swanctl.conf  file is:
--------------------------------------------------

> include conf.d/vcn/*.conf
> secrets {
>     ike {
>         secret = "6rhT;.{s!OgJbkBNa22lFm&#NlCwVk@]"
>     }
> }
> connections {
>     ping {
>         children {
>             ping {
>                 local_ts = 11.5.213.128/32[icmp/2048]
>                 remote_ts = 0.0.0.0/0[icmp/2048]
>                 mode = pass
>                 start_action = start
>             }
>         }
>     }
> }

----------------------------------------------------
in the  path /etc/swanctl/conf.d/vcn/*.conf in node 1 we have actual config file corresondinf to node2 is below:
------------------------------------------------------------
> connections {
>     node2.conf {
>         local_addrs = 11.5.213.128
>         remote_addrs = 11.28.205.128
>         dpd_delay = 30s
>         local {
>             auth = psk
>             # no cert required for psk
>         }
>         remote {
>             auth = psk
>         }
>         children {
>            node2.conf{
>                 local_ts = 11.5.213.0/24
>                 remote_ts = 11.28.205.0/24
>                 mode = tunnel
>                 inactivity = 0s
>                 policies = no
>                 start_action = start
>                 dpd_action = restart
>                 esp_proposals =aes256gcm8-sha256-modp2048
>                 rekey_time=60m
>                 rand_time=6m
>                 interface=ipsec0
>                 mark_in=5
>             }
>         }
>         version = 2
>         proposals =aes256gcm8-sha256-modp2048
>         rekey_time=240m
>         rand_time=24m
>     }
> }
-----------------------------------------------

`ip a` interesting information output :

>  ctrl0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
>     link/ether fe:ff:0b:05:d5:02 brd ff:ff:ff:ff:ff:ff
>     inet 11.5.213.2/8 scope global ctrl0
>        valid_lft forever preferred_lft forever
>     inet6 fe80::fcff:bff:fe05:d502/64 scope link
>        valid_lft forever preferred_lft forever
> 
>  ipsec0@ctrl0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
>     link/ether fe:ff:0b:05:d5:80 brd ff:ff:ff:ff:ff:ff
>     inet 11.5.213.128/32 scope global ipsec0
>        valid_lft forever preferred_lft forever
>     inet6 fe80::fcff:bff:fe05:d580/64 scope link
>        valid_lft forever preferred_lft forever

`ip route ` interesting information output 
11.0.0.0/8 dev ctrl0 proto kernel scope link src 11.5.213.2

`ipsec statusall` cmd output in below file
https://pastebin.pl/view/7f82bc42

swanctl logs in node1 are collected by using the cmd `swanctl --log --pretty` are in below file
https://pastebin.pl/view/54400080

`ip x s l` output in node1:
https://pastebin.pl/view/77b86280

### useful information in node2

swanctl.conf  file is:
--------------------------------------------------

> include conf.d/vcn/*.conf
> secrets {
>     ike {
>         secret = "6rhT;.{s!OgJbkBNa22lFm&#NlCwVk@]"
>     }
> }
> connections {
>     ping {
>         children {
>             ping {
>                 local_ts = 11.28.205.128/32[icmp/2048]
>                 remote_ts = 0.0.0.0/0[icmp/2048]
>                 mode = pass
>                 start_action = start
>             }
>         }
>     }
> }

----------------------------------------------------
in the  path /etc/swanctl/conf.d/vcn/*.conf in node 2 we have actual config file corresondinfg to node1 is below:
------------------------------------------------------------
> connections {
>     dal2-qz2-sr2-rk115-s13-10-12-60-85-ipv4 {
>         local_addrs =  11.28.205.128
>         remote_addrs = 11.5.213.128
>         dpd_delay = 30s
>         local {
>             auth = psk
>             # no cert required for psk
>         }
>         remote {
>             auth = psk
>         }
>         children {
>             dal2-qz2-sr2-rk115-s13-10-12-60-85-ipv4 {
>                 local_ts =  11.28.205.0/24
>                 remote_ts =  11.5.213.0/24
>                 mode = tunnel
>                 inactivity = 0s
>                 policies = no
>                 start_action = start
>                 dpd_action = restart
>                 esp_proposals =aes256gcm8-sha256-modp2048
>                 rekey_time=60m
>                 rand_time=6m
>                 interface=ipsec0
>                 mark_in=5
>             }
>         }
>         version = 2
>         proposals =aes256gcm8-sha256-modp2048
>         rekey_time=240m
>         rand_time=24m
>     }
> }
----------------------------------------------

`ip a` cmd  interesting output in node2 is below
feff0b1ccd02: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default qlen 1000
    link/ether fe:ff:0b:1c:cd:02 brd ff:ff:ff:ff:ff:ff
    inet 11.28.205.2/8 brd 11.255.255.255 scope global feff0b1ccd02
       valid_lft forever preferred_lft forever
    inet6 fe80::fcff:bff:fe1c:cd02/64 scope link 
       valid_lft forever preferred_lft forever

ipsec0@ipsec0-ph: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default qlen 1000
    link/ether fe:ff:0b:1c:cd:80 brd ff:ff:ff:ff:ff:ff
    inet 11.28.205.128/24 brd 11.28.205.255 scope global ipsec0
       valid_lft forever preferred_lft forever
    inet6 fe80::fcff:bff:fe1c:cd80/64 scope link 
       valid_lft forever preferred_lft forever
      
` ip route` cmd interesting output in node 2 is below
 11.0.0.0/8 dev feff0b1ccd02 proto kernel scope link src 11.28.205.2 
11.5.212.128 via 11.28.205.1 dev ipsec0 
11.5.213.128 via 11.28.205.1 dev ipsec0 

note: we are not installing ike-policeies in node1 , in node1 policies are handled by our internal p4 programs
          we are installing states in node 1.
           in node2 we are installing the ike policies and as well as states

**Comment by tobiasbrunner** [comment]
> swanctl logs in node1 are collected by using the cmd `swanctl --log --pretty` are in below file

Please don't. Use the log settings given in #196 for debugging. And do so on both sides to see what's actually going on.

**Reply by trineshreddy** [reply]
hi @tobiasbrunner ,
1)i added the content in the file "/etc/strongswan.d/charon-logging.conf "   as below in node in which strongswan version is less than 5.7.0

charon {
    filelog {
             /var/log/charon_debug.log {
                time_format = %a, %Y-%m-%d, %H:%M:%S
                default = 2
                net = 1
                enc = 1
                asn = 1
                job = 1
                ike_name = yes
                append = no
                flush_line = yes
        }
    }
    syslog {
    }
}
2)swanctl --reload-settings. (executed this cmd)
3)swanctl --load-all (executed this cmd also)
 but unable to see the log file at location "/var/log/charon_debug.log"
please help

**Reply by tobiasbrunner** [reply]
Check your `strongswan.conf` file. This snippet only has an effect if the file contains something like `include /etc/strongswan.d/*.conf`. Otherwise, you can just put this content directly in `strongswan.conf`.

**Reply by trineshreddy** [reply]
` cat /etc/strongswan.conf `
# strongswan.conf - strongSwan configuration file
#
# Refer to the strongswan.conf(5) manpage for details
#
# Configuration changes should be made in the included files

charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
}

include strongswan.d/*.conf
charon {
    filelog {
              /var/log/charon_debug.log {
                time_format = %a, %Y-%m-%d, %H:%M:%S
                default = 2
                net = 1
                enc = 1
                asn = 1
                job = 1
                ike_name = yes
                append = no
                flush_line = yes
        }
    }
    syslog {
}
}

i directly added content in strongswan.conf and executed 2 cmds  `swanctl --reload-settings`. & `swanctl --load-all `
but still i am unable to see logfile created at /var/log/charon_debug.log

> When you create a log file, use the log settings below, unless we tell you otherwise. If you (or your distribution) use a Linux Security Module (LSM), like AppArmor, Selinux, YAMA or TOMOYO, you need to allow the IKE daemon (charon, charon-systemd etc.) to create and write to that file first, or disable the LSM for the time of the debugging. Obviously, allowing the daemon to create and write the file is preferred

here as specified above i donot know if my linux machine use Linux security module(LSM), but anyway how to allow deamon to create and write the file?

**Reply by tobiasbrunner** [reply]
Try a different path (e.g. in `/tmp`), also check the system logs for errors (with the above settings strongSwan will also log via `syslog` so there should be messages in "global" log files).
